File: src/test/cpp/instrumentationTest.cpp
#include "instrumentation_test.hpp"

using std::string;

namespace kotlin_tracer {
TEST(InstrumentationTest, CheckCoroutinesDebugProbesInstrumentationForSwitchTable) {
  auto temp_file_path = runJVMTest("io/github/maximgorbunov/InstrumentationTest.switchTableSuspend",
                                   "io.github.maximgorbunov.InstrumentationTest.simpleFunctionSwitchTableTest");
  checkExecutionTime(temp_file_path, 100000000, 200000000);
  std::filesystem::remove(temp_file_path->string());
}

TEST(InstrumentationTest, CheckCoroutinesDebugProbesInstrumentationForComplexSwitchTable) {
  auto temp_file_path = runJVMTest("io/github/maximgorbunov/InstrumentationTest.switchTableComplexSuspend",
                                   "io.github.maximgorbunov.InstrumentationTest.complexFunctionSwitchTableTest");
  checkExecutionTime(temp_file_path, 200000000, 300000000);
  std::filesystem::remove(temp_file_path->string());
}

TEST(InstrumentationTest, CheckCoroutinesDebugProbesInstrumentationWithoutSwitchTable) {
  auto temp_file_path = runJVMTest("io/github/maximgorbunov/InstrumentationTest.suspendWithoutTable",
                                   "io.github.maximgorbunov.InstrumentationTest.simpleFunctionWithoutSwitchTableTest");
  checkExecutionTime(temp_file_path, 100000000, 200000000);
  std::filesystem::remove(temp_file_path->string());
}

TEST(InstrumentationTest, CheckPlainFunctionInstrumentation) {
  auto temp_file_path = runJVMTest("io/github/maximgorbunov/InstrumentationTest.plainFunction",
                                   "io.github.maximgorbunov.InstrumentationTest.plainFunctionTest");
  checkExecutionTime(temp_file_path, 100000000, 200000000);
  std::filesystem::remove(temp_file_path->string());
}

TEST(InstrumentationTest, CheckDispatchDefualt) {
  auto temp_file_path = runJVMTest("io/github/maximgorbunov/InstrumentationTest.switchTableComplexSuspend",
                                   "io.github.maximgorbunov.InstrumentationTest.dispatchDefaultTest");
  checkExecutionTime(temp_file_path, 200000000, 300000000);
  std::filesystem::remove(temp_file_path->string());
}

TEST(InstrumentationTest, CheckDispatchUndispatched) {
  auto temp_file_path = runJVMTest("io/github/maximgorbunov/InstrumentationTest.switchTableComplexSuspend",
                                   "io.github.maximgorbunov.InstrumentationTest.dispatchUndispatchedTest");
  checkExecutionTime(temp_file_path, 200000000, 300000000);
  std::filesystem::remove(temp_file_path->string());
}

TEST(InstrumentationTest, CheckDispatchUndispatchedWithoutSuspend) {
  auto temp_file_path = runJVMTest("io/github/maximgorbunov/InstrumentationTest.suspendWithoutSuspensionsFunction",
                                   "io.github.maximgorbunov.InstrumentationTest.dispatchUndispatchedWithoutSuspendTest");
  checkExecutionTime(temp_file_path, 100000000, 200000000);
  std::filesystem::remove(temp_file_path->string());
}

TEST(InstrumentationTest, CheckDispatchLazy) {
  auto temp_file_path = runJVMTest("io/github/maximgorbunov/InstrumentationTest.switchTableComplexSuspend",
                                   "io.github.maximgorbunov.InstrumentationTest.dispatchLazyTest");
  checkExecutionTime(temp_file_path, 200000000, 300000000);
  std::filesystem::remove(temp_file_path->string());
}

TEST(InstrumentationTest, CheckDispatchAtomic) {
  auto temp_file_path = runJVMTest("io/github/maximgorbunov/InstrumentationTest.switchTableComplexSuspend",
                                   "io.github.maximgorbunov.InstrumentationTest.dispatchAtomicTest");
  checkExecutionTime(temp_file_path, 200000000, 300000000);
  std::filesystem::remove(temp_file_path->string());
}

}  // namespace kotlin_tracer

File: src/test/kotlin/io.github.maximgorbunov/InstrumentationTest.kt
package io.github.maximgorbunov

import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Test

class InstrumentationTest {

    @Test
    fun simpleFunctionSwitchTableTest(): Unit = runBlocking {
        switchTableSuspend()
    }

    @Test
    fun complexFunctionSwitchTableTest(): Unit = runBlocking {
        switchTableComplexSuspend(2)
    }

    @Test
    fun simpleFunctionWithoutSwitchTableTest(): Unit = runBlocking {
        suspendWithoutTable()
    }

    @Test
    fun dispatchDefaultTest() = runBlocking {
        launch(start = CoroutineStart.DEFAULT) {
            switchTableComplexSuspend(2)
        }.join()
    }

    @Test
    fun dispatchUndispatchedTest() = runBlocking {
        launch(start = CoroutineStart.UNDISPATCHED) {
            switchTableComplexSuspend(2)
        }.join()
    }

    @Test
    fun dispatchUndispatchedWithoutSuspendTest() = runBlocking {
        launch(start = CoroutineStart.UNDISPATCHED) {
            suspendWithoutSuspensionsFunction()
        }.join()
    }

    @Test
    fun dispatchLazyTest() = runBlocking {
        launch(start = CoroutineStart.LAZY) {
            switchTableComplexSuspend(2)
        }.join()
    }

    @Test
    fun dispatchAtomicTest() = runBlocking {
        launch(start = CoroutineStart.ATOMIC) {
            switchTableComplexSuspend(2)
        }.join()
    }

    @Test
    fun plainFunctionTest() {
        plainFunction()
    }

    fun plainFunction() {
        Thread.sleep(100)
    }

    fun suspendWithoutSuspensionsFunction() {
        Thread.sleep(100)
    }


    suspend fun switchTableSuspend() {
        delay(100)
        println("stub")
    }

    suspend fun switchTableComplexSuspend(a: Int): Int {
        delay(100)
        println("stub")
        delay(100)
        if (a > 0) return 2
        else return 3
    }

    suspend fun suspendWithoutTable() {
        delay(100);
    }
}
